waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/service/exchange/ProofOfPossessionParameterFactory.kt

```diff
@@ -0,0 +1,120 @@
+package id.walt.webwallet.service.exchange
+
+import COSE.OneKey
+import com.nimbusds.jose.jwk.ECKey
+import id.walt.crypto.utils.JsonUtils.toJsonElement
+import id.walt.did.dids.DidService
+import id.walt.oid4vc.data.CredentialOffer
+import id.walt.oid4vc.data.OfferedCredential
+import id.walt.oid4vc.data.ProofOfPossession
+import id.walt.oid4vc.data.ProofType
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.encodeToJsonElement
+
+object ProofOfPossessionParameterFactory {
```

Comment:
nitpicking alert!
this could be refactored even more, have a separate method for building the proof-type from each key and did, for cwt and jwt, respectively and also have the instantiation of possesion-parameters extracted into separate methods, e.g.:
```kotlin
private fun cwtProofOfPossessionParameters(it: ProofOfPossession.CWTProofBuilder) =
    ProofOfPossessionParameters(
        ProofType.cwt,
        Json.encodeToJsonElement(it.headers.toCBOR()),
        Json.encodeToJsonElement(it.payload.toCBOR()),
    )
```
and for jwt:
```kotlin
private fun jwtProofOfPossessionParameters(it: ProofOfPossession.JWTProofBuilder) =
    ProofOfPossessionParameters(
        ProofType.jwt,
        it.headers.toJsonElement(),
        it.payload.toJsonElement(),
    )
```
then building the proof type for cwt from key:
```kotlin
private suspend fun cwtProofTypeFromKey(
    credentialOffer: CredentialOffer,
    nonce: String?,
    key: Key
) = let {
    ProofOfPossession.CWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        nonce = nonce,
        coseKey = OneKey(
            ECKey.parse(key.getPublicKey().exportJWK()).toECPublicKey(),
            null
        ).AsCBOR().EncodeToBytes(),
    ).let {
        cwtProofOfPossessionParameters(it)
    }
}
```
and from did:
```kotlin
private fun cwtProofTypeFromDid(credentialOffer: CredentialOffer, did: String, nonce: String?) = let {
    ProofOfPossession.CWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        clientId = did,
        nonce = nonce,
    ).let {
        cwtProofOfPossessionParameters(it)
    }
}
```
and the same thing for jwt:
```kotlin
private suspend fun jwtProofTypeFromKey(
    credentialOffer: CredentialOffer,
    nonce: String?,
    key: Key
) = let {
    ProofOfPossession.JWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        nonce = nonce,
        keyJwk = key.getPublicKey().exportJWKObject(),
        keyId = key.getKeyId(),
    ).let {
        jwtProofOfPossessionParameters(it)
    }
}
```
```kotlin
private fun jwtProofTypeFromDid(
    credentialOffer: CredentialOffer,
    did: String,
    nonce: String?,
    keyId: String
) = let {
    ProofOfPossession.JWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        clientId = did,
        nonce,
        keyId,
    ).let {
        jwtProofOfPossessionParameters(it)
    }
}
```

then `keyProofParameters` would look like:
```kotlin
private suspend fun keyProofParameters(
    did: String,
    offeredCredential: OfferedCredential,
    credentialOffer: CredentialOffer,
    nonce: String?
): ProofOfPossessionParameters {
    val key = DidService.resolveToKey(did).getOrThrow()
    val proofType = offeredCredential.proofTypesSupported?.keys?.first() ?: ProofType.jwt
    return when (proofType) {
        ProofType.cwt -> cwtProofTypeFromKey(credentialOffer, nonce, key)
        ProofType.ldp_vp -> TODO("ldp_vp proof not yet implemented")
        else -> jwtProofTypeFromKey(credentialOffer, nonce, key)
    }
}
```
and the `didProofParameters`:
```kotlin
private fun didProofParameters(
    did: String,
    keyId: String,
    offeredCredential: OfferedCredential,
    credentialOffer: CredentialOffer,
    nonce: String?
): ProofOfPossessionParameters {
    val proofType = offeredCredential.proofTypesSupported?.keys?.first() ?: ProofType.jwt
    return when (proofType) {
        ProofType.cwt -> cwtProofTypeFromDid(credentialOffer, did, nonce)
        ProofType.ldp_vp -> TODO("ldp_vp proof not yet implemented")
        else -> jwtProofTypeFromDid(credentialOffer, did, nonce, keyId)
    }
}
```